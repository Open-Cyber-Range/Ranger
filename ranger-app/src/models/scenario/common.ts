export type Source = {
  name: string;
  version: string;
};
