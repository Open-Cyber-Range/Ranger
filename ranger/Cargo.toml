[package]
name = "ranger"
version = "2.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[package.metadata.deb]
name = "ranger-server"
maintainer = "C14 developers <developers@cr14.ee>"
copyright = "MIT"
depends = "$auto, systemd"
section = "utility"
priority = "optional"
changelog = "changelog"
assets = [
  [
    "target/release/ranger",
    "/var/opt/ranger/bin/",
    "755",
  ],
  [
    "assets/default-production-config.yml",
    "/etc/opt/ranger/ranger/config.yml",
    "644",
  ],
  [
    "../ranger-app/build/assets/*",
    "/etc/opt/ranger/ranger-app/assets",
    "644",
  ],
  [
    "../ranger-app/build/*",
    "/etc/opt/ranger/ranger-app/",
    "644",
  ],
]
systemd-units = { unit-name = "ranger", unit-scripts = "assets/unit-scripts" }

[dependencies]
actix = "0.13"
actix-multipart = "0.6"
actix-web = "4"
actix-web-actors = "4"
actix-web-grants = "3"
actix-http = { version = "3", features = ["http2", "ws"] }
actix-web-httpauth = "0.8"
anyhow = "1"
async-trait = "0.1"
bigdecimal = { version = "0.4", features = ["serde"] }
chrono = { version = "0.4", features = ["serde"] }
diesel = { version = "2.1", features = ["mysql", "r2d2", "chrono", "numeric"] }
diesel_migrations = "2.1"
env_logger = "0.9"
futures = "0.3"
futures-util = "0.3"
jsonwebtoken = "8"
keycloak = { git = "https://github.com/ebakoba/keycloak", branch = "master" }
log = "0.4"
lazy_static = "1.4"
lettre = { version = "0.10", default-features = false, features = [
  "smtp-transport",
  "pool",
  "hostname",
  "builder",
  "rustls",
  "rustls-tls",
] }
sha3 = "0.10"
digest = "0.10"
rand = "0.8"
ranger-grpc = { path = "../ranger-grpc" }
reqwest = { version = "0.11", default-features = false, features = [
  "json",
  "rustls",
  "rustls-tls",
  "tokio-rustls",
  "hyper-rustls",
  "rustls-native-certs",
] }
sdl-parser = "0"
serde = { version = "1.0", features = ["derive"] }
serde_json = "1"
serde_yaml = "0.9"
semver = "1"
tempfile = "3"
thiserror = "1"
toml-query = "0.10"
toml = "0.5.6"
tokio = { version = "1", features = [
  "rt-multi-thread",
  "time",
  "fs",
  "macros",
  "net",
  "rt",
] }
tonic = "0.8"
uuid = { version = "1.1", features = ["v4", "serde", "fast-rng"] }

[dev-dependencies]
actix-rt = "2"
insta = { version = "1", features = ["redactions", "yaml"] }
sdl-parser = { version = "0", features = ["test"] }
