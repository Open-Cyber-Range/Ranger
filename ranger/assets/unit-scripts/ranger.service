[Unit]
Description=ranger
After=network.target

[Service]
ExecStart=/var/opt/ranger/bin/ranger /etc/opt/ranger/ranger/config.yml
Type=simple
Restart=on-failure
RestartSec=10
StartLimitInterval=10m
StartLimitBurst=5

[Install]
WantedBy=multi-user.target
