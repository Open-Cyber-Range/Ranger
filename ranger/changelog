ranger (0.2.0) stretch; urgency=medium

  * Initial Release
  * Adds web application support

 -- Kaarel Allemann <kaarel.allemann@cr14.ee>  Sat, 05 Nov 2022 19:23:51 +0200
