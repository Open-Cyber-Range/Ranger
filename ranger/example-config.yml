host: localhost
port: 8080
repository:
  folder: /tmp/test-repo
  username: some-username
  email: some@email.com
package_folder: /tmp/packages
