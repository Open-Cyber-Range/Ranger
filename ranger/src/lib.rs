pub mod capability;
pub mod configuration;
pub mod database;
pub mod deployers;
pub mod errors;
pub mod machiner;
pub mod node;
pub mod routes;

use crate::database::Database;
use actix::{Actor, Addr};
use anyhow::{anyhow, Result};
use deployers::{get_deployer_capabilities, AddDeployerGroups, DeployerGroups};

use std::collections::HashMap;

pub struct AppState {
    pub database_address: Addr<Database>,
    pub deployer_actor_address: Addr<DeployerGroups>,
}

impl AppState {
    pub fn new() -> Self {
        AppState {
            database_address: Database::new().start(),
            deployer_actor_address: DeployerGroups::new().start(),
        }
    }

    pub async fn add_initial_deployergroups(
        &self,
        deployment_groups: HashMap<String, Vec<String>>,
        deployers: HashMap<String, String>,
        default_deployer_group: String,
    ) -> Result<()> {
        let mut deployer_groups = DeployerGroups::initialize_with_group_names(&deployment_groups);
        let mut deployers = get_deployer_capabilities(deployers).await?;

        deployers.iter_mut().for_each(|deployer| {
            deployment_groups.iter().for_each(|deployer_group| {
                if deployer_group.1.contains(&deployer.deployer_name) {
                    if let Some(deployer_group) = deployer_groups.0.get_mut(deployer_group.0) {
                        deployer_group.insert_by_capability(deployer);
                    }
                }
            });
        });
        let default_deployer_group_value = deployer_groups
            .0
            .get(&default_deployer_group)
            .ok_or_else(|| anyhow!("Default group with given name not found"))?
            .clone();

        deployer_groups
            .0
            .insert("default".to_string(), default_deployer_group_value);

        self.deployer_actor_address
            .send(AddDeployerGroups(deployer_groups))
            .await?;
        Ok(())
    }
}

impl Default for AppState {
    fn default() -> Self {
        Self::new()
    }
}
