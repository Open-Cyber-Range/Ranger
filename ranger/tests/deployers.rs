mod common;

#[cfg(test)]
mod tests {
    use crate::common::create_mock_capability_server;
    use anyhow::Result;
    use ranger::{
        deployers::GetDeployerGroups,
        errors::{RangerError, ServerResponseError},
        AppState,
    };
    use std::collections::HashMap;

    #[actix_web::test]
    async fn add_deployer_groups() -> Result<()> {
        let mut deployers = HashMap::new();
        deployers.insert(
            "my_cool_machiner_one".to_string(),
            format!("http://{}", create_mock_capability_server().run()?),
        );
        deployers.insert(
            "my_cool_machiner_two".to_string(),
            format!("http://{}", create_mock_capability_server().run()?),
        );

        let mut deployer_groups = HashMap::new();
        deployer_groups.insert(
            "my_cool_machiner_group_one".to_string(),
            vec![
                "my_cool_machiner_one".to_string(),
                "my_cool_machiner_two".to_string(),
            ],
        );
        deployer_groups.insert(
            "my_cool_machiner_group_two".to_string(),
            vec!["my_cool_machiner_one".to_string()],
        );

        let app_state = AppState::new();
        app_state
            .add_initial_deployergroups(deployer_groups, deployers)
            .await?;

        let validated_deployer_groups = app_state
            .deployer_actor_address
            .send(GetDeployerGroups)
            .await
            .map_err(|_| ServerResponseError(RangerError::ActixMailBoxError.into()))
            .unwrap();
        insta::with_settings!({sort_maps => true}, {
           insta::assert_yaml_snapshot!(validated_deployer_groups, {
              "[\"my_cool_machiner_group_one\"].machiners[\"my_cool_machiner_one\"]" => "http://some-host:some-port",
              "[\"my_cool_machiner_group_one\"].machiners[\"my_cool_machiner_two\"]" => "http://some-host:some-port",
              "[\"my_cool_machiner_group_two\"].machiners[\"my_cool_machiner_one\"]" => "http://some-host:some-port"
           });
        });
        Ok(())
    }
}
