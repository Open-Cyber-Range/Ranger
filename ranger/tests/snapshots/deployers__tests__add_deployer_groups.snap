---
source: ranger/tests/deployers.rs
expression: validated_deployer_groups
---
default:
  machiners:
    my_cool_machiner_one: "http://some-host:some-port"
    my_cool_machiner_two: "http://some-host:some-port"
  switchers: {}
  templaters: {}
my_cool_machiner_group_one:
  machiners:
    my_cool_machiner_one: "http://some-host:some-port"
    my_cool_machiner_two: "http://some-host:some-port"
  switchers: {}
  templaters: {}
my_cool_machiner_group_two:
  machiners:
    my_cool_machiner_one: "http://some-host:some-port"
  switchers: {}
  templaters: {}

